File: kotlin/knapsack/src/main/kotlin/Knapsack.kt
data class Item(val weight: Int, val value: Int)

fun knapsack(maximumWeight: Int, items: List<Item>): Int {
    val n = items.size
    val dp = Array(n + 1) { IntArray(maximumWeight + 1) }

    for (i in 1..n) {
        val (weight, value) = items[i - 1]

        for (w in 0..maximumWeight) {
            dp[i][w] = dp[i - 1][w]

            if (weight <= w) {
                dp[i][w] = maxOf(dp[i][w], dp[i - 1][w - weight] + value)
            }
        }
    }

    return dp[n][maximumWeight]
}

File: cpp/knapsack/knapsack.cpp
#include "knapsack.h"

namespace knapsack
{

    int maximum_value(int max_weight, const std::vector<knapsack::Item> &items)
    {
        int n = static_cast<int>(items.size());
        std::vector<std::vector<int>> dp(n + 1, std::vector<int>(max_weight + 1, 0));

        for (int i = 1; i <= n; i++)
        {
            int weight = items[i - 1].weight;
            int value = items[i - 1].value;

            for (int w = 0; w <= max_weight; w++)
            {
                dp[i][w] = dp[i - 1][w];

                if (weight <= w)
                {
                    dp[i][w] = std::max(dp[i][w], dp[i - 1][w - weight] + value);
                }
            }
        }

        return dp[n][max_weight];
    }

} // namespace knapsack
